lib/Performance/Probability.pm

```diff
@@ -0,0 +1,233 @@
+                $start_i = $start_time->[$i];
+                $start_j = $start_time->[$j];
+                $sell_i  = $sell_time->[$i];
+                $sell_j  = $sell_time->[$j];
+
+                if ($start_j > $start_i and $start_j < $sell_i) {
+                    #calculate a, b and c interval. please see the documentation for details
+                    #of a, b and c.
+                    my $a_interval = $start_j - $start_i;
+                    my $b_interval = $sell_i - $start_j;
+                    my $c_interval = $sell_j - $sell_i;
+
+                    if ($c_interval < 0) {
+                        $c_interval = 0 - $c_interval;
+                        $b_interval = $sell_j - $start_i;
+                    }
+
+                    my $i_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$i]);
+                    my $j_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$j]);
+
+                    my $corr_ij = $b_interval / (sqrt($a_interval + $b_interval) * sqrt($b_interval + $c_interval));
+
+                    if ($types->[$i] ne $types->[$j]) {
+                        $corr_ij = -1 * $corr_ij;
+                    }
+
+                    my $p_ij = Math::BivariateCDF::bivnor($i_strike, $j_strike, $corr_ij);
+
+                    my $covariance_ij =
+                        ($p_ij - $pk->[$i] * $pk->[$j]) * ($wk->[$i] - $lk->[$i]) * ($wk->[$j] - $lk->[$j]);
+
+                    $covariance = $covariance + $covariance_ij;
+                }
+            }
+        }
+    }
+
+    return $covariance;
+}
+
+=item B<get_performance_probability>
+
+Calculate performance probability ( modified sharpe ratio );
+
+=cut
+
+sub get_performance_probability {
+
+    my $params = shift;
+
+    my $pnl = $params->{pnl};
+
+    if (not defined $pnl) {
+        die "pnl is a required parameter.";
+    }
+
+    #Below variables are all arrays.
+    my $start_time   = $params->{start_time};
+    my $sell_time    = $params->{sell_time};
+    my $types        = $params->{types};
+    my $underlying   = $params->{underlying};
+    my $bought_price = $params->{bought_price};
+    my $payout       = $params->{payout};
+
+    if (grep { $_ != scalar(@$start_time) } (scalar(@$sell_time), scalar(@$types), scalar(@$underlying), scalar(@$bought_price), scalar(@$payout))) {
+        die "start_time, sell_time, types, underlying, bought_price and payout are required parameters and need to be arrays of same lengths.";
+    }
+
+    my $pk = _build_pk($bought_price, $payout);
+    my $lk = _build_lk($bought_price);
+    my $wk = _build_wk($bought_price, $payout);
+
+    my $mean = _mean($pk, $lk, $wk);
+
+    my $variance = _variance_x_square($pk, $lk, $wk);
+
+    my $covariance = _covariance($start_time, $sell_time, $underlying, $types, $pk, $lk, $wk);
+
+    #Calculate the performance probability here.
+    my $prob = 0;
+
+    $prob = $pnl - $mean;
+    $prob = $prob / sqrt(($variance - ($mean**2.0)) + 2.0 * $covariance);
```

Comment:
Can we make sure, that division by zero can't occur here ? May be you should ask Jeff about pre-conditions for the function?